### ISSUE TITLE
"Add to homescreen" does not create WebAPK

### ISSUE THREAD

**Issue by cassshh** [brave/brave-browser#7357]
**Did you search for similar issues before submitting this one?**
Yes

**Description:**
When adding a PWA to your homescreen it adds a shortcut with the Brave-icon badge.
This behavior is similar to other browsers Edge/Firefox/etc.
However, Chrome creates a WebAPK. This results in the PWA added to your homescreen without a badge; in the app drawer; listed in settings>apps; opens PWA when clicking on corresponding URL.
And in Brave its code, it does seem to have WebAPKs like Chrome has:
https://github.com/brave/browser-android-tabs/tree/master/chrome/android/webapk

**Device Details:**
   - Install Type(ARM, x86): ARM (if i stand correct?)
   - Device(Phone, Tablet, Phablet): Huawei Nexus 6P
   - Android Version: 8.1.0
    
**Brave Version:**
Brave 1.0.48, Chromium 66.0.3359.158

**Steps to reproduce:**
    1. Open a PWA page; e.g. https://pwa.rocks/
    2. Add to homescreen

**Actual Behavior**
Not installed as WebAPK

**Expected Behavior**
Being installed as WebAPK

**Is this an issue with Beta build?**
No

**Is this an issue in the currently released version?**
Yes

**Can this issue be consistently reproduced?**
Yes

**Website problems only:**
  - did you check with Brave Shields down? Yes
  - did you check in Chrome for same behavior? Yes (Chrome 66.0.3359.158)

**Screenshot if needed:**
https://imgur.com/a/h7ikS1t

**Any related issues:**
N/A

**Comment by FingerlessGlov3s** [comment]
In January 2020, I was able to add TheLounge which is a selfhosted web IRC client "https://thelounge.chat/" as a web app to the home screen. I'd just hit the menu button while using the website, press add to home screen and it would become an app.

Sometime this month of February 2020, this no longer works, so there must of been an update which has broken it.
I can confirm the website still supports webapp because I've not updated my instance and Firefox still works as an WebApp.

I've tried clearing data,cache and reinstalling.

Is this a new bug or same problem?

**Comment by Brave-Matt** [comment]
+1 from Community:
https://community.brave.com/t/progressive-web-apps/110241/6

**Comment by szaimen** [comment]
Is there anything new regarding this bug since the migration to brave-core should already be done?

**Comment by immanuelfodor** [comment]
Does this issue related anyhow to the Web Share Target API? I can't share "into" an app that is "installed" on the home screen by Brave even if it is a valid PWA with the related configuration in place. (https://github.com/miniflux/v2/pull/936)

**Comment by nghkhoi** [comment]
> Does this issue related anyhow to the Web Share Target API? I can't share "into" an app that is "installed" on the home screen by Brave even if it is a valid PWA with the related configuration in place. ([miniflux/v2#936](https://github.com/miniflux/v2/pull/936))

Same experience here from a slightly different environment: building a simple Web Share Target PWA and tested on Chrome and Edge on Android, only the PWA installed on Chrome (Android) properly appeared on the native Share UI, the Edge's PWA was nothing more than a shortcut. Apparantly it need to be WebAPK to truly support Web Share Target API, IMHO.

**Comment by ghostdevv** [comment]
We really need this, there is such a difference between installing a PWA via Brave and Chrome. Unfortunately Chrome is one upping Brave here

**Comment by juanferrer** [comment]
Any progress on this?

**Comment by srirambv** [comment]
WebAPK is supported natively on chrome. Need to investigate why its not working on Brave and only supports adds to homescreen.

cc: @anthonypkeane @SergeyZhukovsky

**Comment by anthonypkeane** [comment]
Some more user feedback: https://twitter.com/lukeshiru/status/1486821825799614467?s=20&t=v6XexhvwY9o5JXTqCzyh9A

**Comment by anthonypkeane** [comment]
Linked to https://github.com/brave/brave-browser/issues/6412

**Comment by srirambv** [comment]
+1 from @szaimen via #6412

**Comment by tennox** [comment]
I think android support for doing this via browsers other than chrome is still unavailable, check the last item `[on the dev page](https://developers.google.com/web/fundamentals/integration/webapks)`:
> **I am a developer of another browser on Android, can I have this seamless install process?**
> We are working on it. We are committed to making this available to all browsers on Android and we will have more details soon. 

So it's google to blame, if anyone :tipping_hand_man:

**Comment by fallaciousreasoning** [comment]
This is the related bug in Chromium https://bugs.chromium.org/p/chromium/issues/detail?id=1243583, where Google is tracking interest in the feature being implemented.

**Comment by ygoe** [comment]
@fallaciousreasoning Cross-linked it.

**Comment by dancojocaru2000** [comment]
Would it be possible for Brave to have its own WebAPK server as described in that Chromium bug tracker?

I think the inconvenience of manually installing the APK isn't that big.

**Comment by NicholasFlamy** [comment]
Has anyone revisited this? For me it's not a critical need, but there are some things, like the Chick-Fil-A app, that I feel are pointless to have installed as a full blown app (and has to be updated all the time) when I could just have a lightweight webpage basically.

**Comment by NicholasFlamy** [comment]
> One of the devs [provided this update on X](https://x.com/SergZh_/status/1877754047215669704):
> 
> > I don't think we can do much in Brave currently. https://issues.chromium.org/issues/40195497?pli=1 there is a dep on store(Google's or Samsung's). We don't plan any store in the near feature. I would suggest to leave more comments in the chromium issue to open the server for all chromium based browsers.
> 
> So it might be time to resolve this one and similar and just point people to the chromium issues to comment and lobby for this.

Thank you for this information. Time to add to my list of F*ck Google.

**Comment by aryankaran** [comment]
wouldn't it be possible, if you generate apk on the fly and prompt use for installation alongside reason they have to do, atleast in dev / canary build it could be tried for na ?? then maybe in stable or play builds too ..?

P..S.: I dunno much about internals, justa thought, so forgive for any misunderstanding...✌🏻🙂

**Comment by NicholasFlamy** [comment]
> wouldn't it be possible, if you generate apk on the fly and prompt use for installation alongside reason they have to do, atleast in dev / canary build it could be tried for na ?? then maybe in stable or play builds too ..?
> 
> P..S.: I dunno much about internals, justa thought, so forgive for any misunderstanding...✌🏻🙂

~I don't think webapks are actual apps/apks. They're some really weird in-between.~
Take that with a grain of salt. I haven't looked that deep into them because I figured it wasn't worth my time if they're gatekept anyway.

Edit: Nevermind, they are actual apps/apks.

**Comment by MarckFp** [comment]
> wouldn't it be possible, if you generate apk on the fly and prompt use for installation alongside reason they have to do, atleast in dev / canary build it could be tried for na ?? then maybe in stable or play builds too ..?
> 
> P..S.: I dunno much about internals, justa thought, so forgive for any misunderstanding...✌🏻🙂

I think the only approach on doing this if it's brave underneath uses [bubblewrap cli](https://github.com/GoogleChromeLabs/bubblewrap) or a similar tool to convert a PWA into APK. But it'll still have the concern of how is going to sign the apk to be verified if it's not the owner of the PWA itself. Also since is a CLI tool and takes some resources this would need to run on Brave's servers because Android devices are very limited.

So in the end we just need Google to delete chrome dependencies on webapks and we're screwed with that.

**Comment by fallaciousreasoning** [comment]
Yeah, the other thing is I don't think we could use vanilla bubble wrap, as we'd need to add our own not-quite-trusted-web-activity, as by default TrustedWebActivities will show a custom tab bar is you haven't set up the assetlinks properly

**Comment by aryankaran** [comment]
> > wouldn't it be possible, if you generate apk on the fly and prompt use for installation alongside reason they have to do, atleast in dev / canary build it could be tried for na ?? then maybe in stable or play builds too ..?
> > P..S.: I dunno much about internals, justa thought, so forgive for any misunderstanding...✌🏻🙂
> 
> I think the only approach on doing this if it's brave underneath uses [bubblewrap cli](https://github.com/GoogleChromeLabs/bubblewrap) or a similar tool to convert a PWA into APK. But it'll still have the concern of how is going to sign the apk to be verified if it's not the owner of the PWA itself. Also since is a CLI tool and takes some resources this would need to run on Brave's servers because Android devices are very limited.
> 
> So in the end we just need Google to delete chrome dependencies on webapks and we're screwed with that.

regarding that, my suggestion is multiple signing keys, say a randomly generated key handed over to user (or maybe user provides one whatever) is added to verification chain for pwa verification (I'd there exists sucha thing 😅 idk ) , and which only checks for the source, then a wrapper as you said **bubble wrap** could be used with that signing key provided , and this might solve the primary issue ig ... 🙂🤠

**Comment by adhocore** [comment]
long time brave user here along side other browsers. recently moved to using brave as main+default browser in android; and while all else were great, not having a proper webapk for pwa was kinda bittersweet. it's also amazing that issue remains since 2018, for 8 years now.

**Comment by NicholasFlamy** [comment]
> long time brave user here along side other browsers. recently moved to using brave as main+default browser in android; and while all else were great, not having a proper webapk for pwa was kinda bittersweet. it's also amazing that issue remains since 2018, for 8 years now.

Well, blame Google. The solutions to work around that are a lot of work for a relatively niche feature (I love the feature and think it should go mainstream to cut down on unnecessary apps, but it hasn't happened yet). Unless someone contributes this feature I doubt Brave devs have it that far up the priority list.
